File: 백준/Gold/2467. 용액/용액.java
import java.util.*;
import java.io.*;

public class Main
{
	public static void main(String[] args) throws IOException {
		final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		final int N = Integer.parseInt(br.readLine());
		final int[] nums = new int[N];
		
		final StringTokenizer st = new StringTokenizer(br.readLine()," ");
		
		for (int i = 0; i < N; i++) {
		    nums[i] = Integer.parseInt(st.nextToken());
		}
		
		int start = 0;
		int end = N-1;
		int min_blending = Math.abs(nums[start] + nums[end]);
		
		final int[] answer = new int[] {nums[start], nums[end]};
		
		while (start < end) {
		    final int blending = nums[start] + nums[end];
		    
		    if (min_blending > Math.abs(blending)) {
		        min_blending = Math.abs(blending);
		        answer[0] = nums[start];
		        answer[1] = nums[end];
		    }
		    
		    if (blending < 0) {
		        start++;
		    } else {
		        end--;
		    }
		}
		System.out.println(answer[0] + " " + answer[1]);
	}
}

File: 백준/Gold/2467. 용액/용액.py
from sys import stdin

input = stdin.readline
n = int(input().strip())
nums = list(map(int, input().strip().split()))

def bin_search(start, end):
    answer = nums[start], nums[end]
    max_blending = abs(nums[start] + nums[end])

    while start < end:
        blending = nums[start] + nums[end]

        if abs(blending) < max_blending:
            max_blending = abs(blending)
            answer = nums[start], nums[end]
    
        if blending < 0:
            start += 1
        else:
            end -= 1
            
    return answer
    

print(*bin_search(0, n-1))
